### PR TITLE
[c++] Fix sanitizer flags and move from an option to a config type

### PR DESCRIPTION
**Pull request by jp-dark**
The builds using the `SANITIZER` option were missing the linker flags necessary to build. This PR adds those back in and moves enabling sanitizers to a config option. The new build types are enabled:

* `ASAN`: address sanitizer 
* `TSAN`: thread sanitizer
* `LSAN`: leak sanitizer
* `MSAN`: memory sanitizer
* `UBSAN`: undefined-behavior sanitizer

### CODE REVIEW

**Review by johnkerl — COMMENTED** [review]
LGTM as long as the sanitizer is sanitizing

I defer to @nguyenv here ...

**Review by johnkerl — APPROVED** [review]
LGTM as long as the sanitizer is sanitizing

I defer to @nguyenv here ...

**Review by nguyenv — APPROVED** [review]

File: libtiledbsoma/cmake/Superbuild.cmake
Comment (teo-tsirpanis) [inline libtiledbsoma/cmake/Superbuild.cmake:66]:
```suggestion
```

We don't need this with the multi-config arrangement, right?

File: libtiledbsoma/CMakeLists.txt
Comment (teo-tsirpanis) [inline libtiledbsoma/CMakeLists.txt:95]:
```suggestion
  set(CMAKE_CONFIGURATION_TYPES
      "Release;Debug;RelWithDebInfo;ASAN;TSAN;LSAN;UBSAN;MSAN"
      CACHE
      STRING
      "Semi-colon separate list of build types for multi-configuration generators."
  )
  set(CMAKE_MAP_IMPORTED_CONFIG_ASAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_TSAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_LSAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_UBSAN Debug)
  set(CMAKE_MAP_IMPORTED_CONFIG_MSAN Debug)
```

File: libtiledbsoma/CMakeLists.txt
Comment (teo-tsirpanis) [inline libtiledbsoma/CMakeLists.txt:69]:
```suggestion
```

File: libtiledbsoma/src/CMakeLists.txt
Comment (teo-tsirpanis) [inline libtiledbsoma/src/CMakeLists.txt:261]: This should better go to `target_link_options`, but `libraries` works as well.

**Review by teo-tsirpanis — COMMENTED** [review]
It might have been cleaner if the sanitizer was an option separate from the config, and either accept that it will work only in Debug mode, or enforce it by setting `CMAKE_CONFIGURATION_TYPES` to only `Debug`.

File: libtiledbsoma/cmake/Superbuild.cmake
Comment (jp-dark) [reply libtiledbsoma/cmake/Superbuild.cmake:66]: Enabling the sanitizer on TileDB embedded is a separate option.

Comment (teo-tsirpanis) [reply libtiledbsoma/cmake/Superbuild.cmake:66]: Oh I forgot about that.

File: libtiledbsoma/CMakeLists.txt
Comment (jp-dark) [reply libtiledbsoma/CMakeLists.txt]: Thanks!

**Comment by jp-dark** [comment]
[sc-53780](https://app.shortcut.com/tiledb-inc/story/53780/fix-address-sanitizer-in-libtiledbsoma-cmake)

**Pull request merged**
